src/front/glsl_new/parser.rs

```diff
@@ -130,37 +130,40 @@ pomelo! {
 
     // expression
     variable_identifier ::= Identifier(v) {
-        if v.1.as_str() == "gl_Position" &&
-            (extra.shader_stage == ShaderStage::Vertex ||
-            extra.shader_stage == ShaderStage::Fragment) {
-            let h = extra.global_variables.fetch_or_append(
-                GlobalVariable {
-                    name: Some(v.1.clone()),
-                    class: match extra.shader_stage {
-                        ShaderStage::Vertex => StorageClass::Output,
-                        ShaderStage::Fragment => StorageClass::Input,
-                        _ => StorageClass::Input,
-                    },
-                    binding: Some(Binding::BuiltIn(BuiltIn::Position)),
-                    ty: extra.types.fetch_or_append(Type {
-                        name: None,
-                        inner: TypeInner::Vector {
-                            size: VectorSize::Quad,
-                            kind: ScalarKind::Float,
-                            width: 4,
+        if v.1.as_str() == "gl_Position" {
+            match extra.shader_stage {
+            ShaderStage::Vertex | ShaderStage::Fragment { .. } => {
```

Comment:
Not sure if it matters, but like this, a variable named `gl_Position` will no longer be allowed at all for non vertex/fragment stages.

Reply:
Please explain further. I was only fixing compilation errors, not logic.

The original already prohibited it everywhere else, no?

Reply:
No, the old version would treat `gl_Position` as a normal variable for stages other than vertex/fragment - it was all tested for in the same `if`.

Reply:
In your version, this would go through the default arm of the `match`, returning `ParseFail` error

Reply:
Again not sure how much it matters - dunno if people e.g. write compute shaders with vars named `gl_Position` :D

Reply:
You meant to comment the error then :). The solution is to simply remove the ParseFail. I can do that.

Reply:
Oh and I need to reorder the if and the match :/

Reply:
not sure that will be the same ... it needs to go to the else of the if, where it treats it like a normal var

Reply:
that's why before it was all tested in the same condition in the if :)

Reply:
Sadly, Rust still doesn't allows for multiple if let's and ifs in a single statement :(.

Reply:
I tried fixing it this way:
```
    variable_identifier ::= Identifier(v) {
        match extra.shader_stage {
        ShaderStage::Vertex | ShaderStage::Fragment { .. } => {
            if v.1.as_str() == "gl_Position" {
                // ...
                return ExpressionRule {
                    expression,
                    statements: vec![],
                };
            }
        },
        _ => {},
        };

        let expression =
        if let Some(local_var) = extra.context.lookup_local_var(&v.1) {
            local_var
        } else if let Some(global_var) = extra.context.lookup_global_var_exps.get(&v.1) {
            *global_var
        } else {
            return Err(ErrorKind::UnknownVariable(v.0, v.1))
        };

        ExpressionRule {
            expression,
            statements: vec![],
        }
    }
```

But that just gives me note: 
expected enum `std::result::Result<(), front::glsl_new::error::ErrorKind>`
found struct `front::glsl_new::ast::ExpressionRule`

Which is weird, since It was not result and making it Ok just gives error that is it not ().

I don't understand this parser at all. Can you give me a hand?

Reply:
Yeah, think I've been fighting with similar issues about `if let` vs. `match` - I'm still a Rust noob 😄 

In the pomelo parser you should be able to return either the type of the rule (here `ExpressionRule`) or an error reulst. Not sure what is causing trouble here :S

Reply:
If it helps, feel free to move all this logic out to e.g. `variables.rs`, been meaning to refactor logic out of `parser.rs` anyway, to keep it small and readable.